File: ast/statement/switch.go
package statement

import (
	"github.com/panda-io/micro-panda/ast"
	"github.com/panda-io/micro-panda/ast/core"
	"github.com/panda-io/micro-panda/ast/types"
	"github.com/panda-io/micro-panda/token"
)

type Switch struct {
	StatementBase
	Initialization core.Statement
	Operand        ast.Expression
	Cases          []*Case
	Default        *Case
}

type Case struct {
	StatementBase
	Token token.Token
	Case  ast.Expression
	Body  core.Statement
}

func (s *Switch) Validate(c ast.Context) {
	ctx := c.NewContext()
	if s.Initialization != nil {
		s.Initialization.Validate(ctx)
	}
	var operandType core.Type
	if s.Operand == nil {
		c.Error(s.GetPosition(), "expect switch operand")
		return
	} else {
		s.Operand.Validate(ctx, nil)
		operandType = s.Operand.Type()
		if !types.IsInteger(operandType) {
			c.Error(s.Operand.GetPosition(), "expect integer operand")
			return
		}
	}
	for _, ca := range s.Cases {
		caseCtx := ctx.NewContext()
		ca.Validate(caseCtx, operandType)
	}
	if s.Default != nil {
		defaultCtx := ctx.NewContext()
		s.Default.Validate(defaultCtx, operandType)
	}
}

func (c *Case) Validate(ctx core.Context, operandType core.Type) {
	if c.Case == nil {
		ctx.Error(c.GetPosition(), "expect case expression")
	} else {
		c.Case.Validate(ctx, operandType)
		if !c.Case.Type().Equal(operandType) {
			ctx.Error(c.GetPosition(), "case operand type mismatch with switch operand")
		}
	}
	if c.Body != nil {
		c.Body.Validate(ctx)
	}
}

File: MicroPanda/AST/Statement/Switch.cs
namespace MicroPanda.AST.Statement;

using Expression;
using Node;
using Token;
using Type;

internal class Switch : Statement
{
    internal Statement? Initialization { get; set; }
    internal Expression? Operand { get; set; }
    internal List<Case> Cases = [];
    internal Case? Default { get; set; }

    internal override void Validate(Context context)
    {
        var ctx = context.NewContext();
        Initialization?.Validate(ctx);
        Type operandType;
        if (Operand == null)
        {
            context.Program.Error(Position, "expect switch operand");
            return;
        }
        else
        {
            Operand.Validate(ctx, null);
            operandType = Operand.Type!;
            if (!TypeHelper.IsInteger(operandType))
            {
                context.Program.Error(Operand.Position, "expect integer operand");
                return;
            }
        }
        foreach (var ca in Cases)
        {
            var caseCtx = ctx.NewContext();
            ca.Validate(caseCtx, operandType);
        }
        if (Default != null)
        {
            var defaultCtx = ctx.NewContext();
            Default.Validate(defaultCtx, operandType);
        }
    }
}

internal class Case : Node
{
    internal Token Token { get; set; }
    internal Expression? CaseExpression { get; set; }
    internal Statement? Body { get; set; }

    internal void Validate(Context context, Type operandType)
    {
        if (CaseExpression == null)
        {
            context.Program.Error(Position, "expect case expression");
        }
        else
        {
            CaseExpression.Validate(context, operandType);
            if (!CaseExpression.Type!.Equal(operandType))
            {
                context.Program.Error(Position, "case operand type mismatch with switch operand");
            }
        }
        Body?.Validate(context);
    }
}